tools/manifest/manifest.py

```diff
@@ -285,119 +155,107 @@ def update(self, tree):
 
-                is_new = rel_path not in path_hash  # type: bool
+                is_new = rel_path_parts not in deleted  # type: bool
                 hash_changed = False  # type: bool
 
                 if not is_new:
-                    old_hash, old_type = path_hash[rel_path]
+                    deleted.remove(rel_path_parts)
+                    old_type = types[rel_path_parts]
+                    old_hash = data[old_type].hashes[rel_path_parts]
+                    file_hash = source_file.hash  # type: Text
                     if old_hash != file_hash:
                         hash_changed = True
+                        del data[old_type][rel_path_parts]
 
                 if is_new or hash_changed:
-                    new_type, manifest_items = source_file.manifest_items()
-                    data[new_type][rel_path] = set(manifest_items)
-                    path_hash[rel_path] = (file_hash, new_type)
-                    if hash_changed and new_type != old_type:
-                        del data[old_type][rel_path]
-                    changed = True
-
-        deleted = prev_files - seen_files
+                    to_update.append(source_file)
+
+        if to_update:
+            changed = True
+
+        if len(to_update) > 25 and cpu_count() > 1:
+            pool = Pool()
+            results = pool.imap_unordered(compute_manifest_items,
+                                          to_update,
+                                          chunksize=max(1, len(to_update) // 10000)
+                                          )  # type: Iterator[Tuple[Tuple[Text, ...], Text, Set[ManifestItem], Text]]
+        elif PY3:
+            results = map(compute_manifest_items, to_update)
+        else:
+            results = itertools.imap(compute_manifest_items, to_update)
+
+        for result in results:
+            rel_path_parts, new_type, manifest_items, file_hash = result
+            data[new_type][rel_path_parts] = manifest_items
+            data[new_type].hashes[rel_path_parts] = file_hash
+
         if deleted:
             changed = True
-            for rel_path in deleted:
-                if rel_path in path_hash:
-                    _, old_type = path_hash[rel_path]
-                    del path_hash[rel_path]
-                    try:
-                        del data[old_type][rel_path]
-                    except KeyError:
-                        pass
-                else:
-                    for test_data in itervalues(data):
-                        if rel_path in test_data:
-                            del test_data[rel_path]
+            for rel_path_parts in deleted:
+                for test_data in itervalues(data):
+                    if rel_path_parts in test_data:
+                        del test_data[rel_path_parts]
 
         return changed
 
-    def to_json(self):
-        # type: () -> Dict[Text, Any]
+    def to_json(self, super_dangerous_mode_do_not_use=False):
```

Comment:
This transfers ownership in really annoying and subtle ways and breaks a lot our tests, but at the same time we need this in our `write` case where it _is_ safe (because no reference to the JSON object outlives the `write` call), though it is only safe as long as it is single-threaded (but I think lots of `TypeData` isn't thread-safe :/)

Reply:
Basically, you can't safely do `Manifest.from_json(m.to_json())` without these optional default-off arguments because for the quick case we really want the callee of both functions to own the first argument/return value so we can avoid copying the entire JSON data structure.

I could call the argument here `caller_owns_return_value=True`, I guess?

Reply:
Yeah, so I think I understand the semantics, but having the option marked as "dangreous" whilst we actually use it seems poor. We should document the invariants that have to be upheld and give the parameter a name that reflects the actual behaviour.